src/callgrind.rs

```diff
@@ -44,82 +68,143 @@ impl Locations {
 
+    // Add to our stats the exclusive time for a given function.
     fn add_exclusive(&mut self, entry: &StackEntry) {
         self.location(&entry.frame).exclusive += entry.exclusive;
     }
 
+    // Add to our stats info about a single call from a parent to a child
+    // function.
     fn add_inclusive(&mut self, parent: &StackFrame, child: &StackEntry) {
         let ploc = self.location(parent);
+        // If we've never seen this parent-child relationship, insert an empty
+        // record.
         let val = ploc.calls.entry(child.frame.clone()).or_insert(Call {
             count: 0,
             inclusive: 0,
         });
+
+        // Add both the count and the inclusive samples cound.
         val.count += 1;
         val.inclusive += child.inclusive;
     }
 }
 
 impl Stats {
+    // Create an empty stats tracker.
     pub fn new() -> Stats {
         Stats {
             ..Default::default()
         }
     }
 
+    // Add a single stack sample to this Stats.
     pub fn add(&mut self, stack: &Vec<StackFrame>) {
-        // We get input with the root of the callgraph at the end. Reverse that!
+        // The input sample has the root function at the end. Reverse that!
         let rev: Vec<_> = stack.iter().rev().collect();
 
-        // Skip any common items
+        // At this point, the previous stack (self.stack) and the new stack
+        // (rev) may have some parts that agree and others that differ:
+        //
+        // Old stack                      New stack
+        // +-------+          ^           +------+
+        // | root  |          |           | root |
+        // +-------+          |           +------+
+        // |   A   |        Common        |   A  |
+        // +-------+          |           +------+
+        // |   B   |          |           |   B  |
+        // +-------+      ^   v    ^      +------+
+        // |   C   |      |        |      |   X  |
+        // +-------+      |     Only new  +------+
+        // |   D   |   Only old    |      |   Y  |
+        // +-------+      |        v      +------+
+        // |   E   |      |
+        // +-------+      v
+        //
+        // Three sections are important:
+        //
+        // 1. The common base (root, A,  B)
+        // 2. The calls only on the previous stack (C, D, E)
+        // 3. The calls only on the new stack (X, Y)
+
+        // 1. Common items we can ignore. Find out how many there are, so we
+        // can skip them.
         let mut common = 0;
         let max_common = min(self.stack.len(), rev.len());
         while common < max_common && &self.stack[common].frame == rev[common] {
             common += 1;
         }
 
-        // Pop old items
+        // 2. Items only on the previous stack won't be kept, so we have to
+        // integrate them into our stats.
         while self.stack.len() > common {
+            // For each entry, pop it from our stored stack, and track its
+            // exclusive sample count.
```

Comment:
I don't understand why **all** of the entries in the stack get their exclusive count updated. From the definition of exclusive above  I thought only the top thing on the stack (`E` in the diagram above) would get its exclusive count incremented, because we can't know that any time was spent in `C` or `D`

exclusive definition I read:

> How many samples were found inside this function only, not including calls to sub-functions?

Reply:
I figured it eventually -- the magic happens down on line 174.

Reply:
Yeah, I should call out that there are two super-steps here:

A. Reconcile the previous and current stack (includes all of 1, 2, 3)
B. Actually count samples